bugtool/cmd/root.go

```diff
@@ -413,6 +427,48 @@ func split(prompt string) (string, []string) {
 	return cmd, args
 }
 
+// hashEncryptionKeys process the buffer containing output of `ip -s xfrm state`.
+// It searches for IPsecKeys in the output and replaces them by its hash.
+func hashEncryptionKeys(output string) string {
+	var finalOutput []string
+	// Split the result as a slice of strings.
+	lines := strings.Split(output, "\n")
+	// Search for lines containing encryption keys.
+	// Note that `auth-trunc` is also a relevant pattern, but we already match on
+	// the more generic `auth` pattern.
+	substrings := [...]string{"auth", "enc", "aead", "comp"}
+	for _, line := range lines {
+		for _, substr := range substrings {
+			r, _ := regexp.Compile(substr)
+			matched := r.MatchString(line)
+			if matched {
+				for i, splits := range strings.Fields(line) {
+					hexFound, _ := regexp.MatchString("0x", splits)
+					if hexFound {
+						splits := strings.Fields(line)
```

Comment:
I had a comment on this in my previous review, it looks like I failed to click to add it to the review and it got lost T_T.

It was about avoiding the additional split on this line. If we can get the index and length of the matched pattern when looking for the hexadecimal string, we could avoid splitting again here and trying to match on each word of the line.

I had a quick look at the functions in the regexp package, and it looks like `FindStringSubmatchIndex()` could be used to retrieve the end and start offsets at the same time as we look for the `auth` or other similar patterns. You should have a look at it, https://golang.org/pkg/regexp/#Regexp.FindStringSubmatchIndex. Here is a sample I did to check how it works:

```go
package main

import "regexp"
import "fmt"

func main() {
	r, _ := regexp.Compile("auth(.*[[:blank:]](0[xX][[:xdigit:]]+))?")

	var testStrings = []string{
		// `auth` and hexa string
		"<garbage> auth foo bar 0x123456af baz",
		// `auth` but no hexa string
		"<garbage> auth foo bar ###23456af baz",
		// nothing
		"<garbage> xxxx foo bar 0x123456af baz",
	}
	for _, s := range testStrings {
		fmt.Printf("%s:\t%v\n", s, r.FindStringSubmatchIndex(s))
	}
}

```
```
$ go run ./prog.go
<garbage> auth foo bar 0x123456af baz:  [10 33 14 33 23 33]
<garbage> auth foo bar ###23456af baz:  [10 14 -1 -1 -1 -1]
<garbage> xxxx foo bar 0x123456af baz:  []
```

Reply:
@qmonnet Thanks for this code snippet. Learned a lot about regex, Also I'll try to fix the failing test.